File: src/io/prompt_handler.nim
import models/prompt_state
import core/session_state
import core/get_prompt
import io/handler_types
import io/handler_registry
import io/read_raw_input
import io/flush

proc handleInput*(state: SessionState): string =
  var prompt = PromptState(
    prompt: getPrompt(state),
    content: "",
    cursorPosition: 0,
    continueReading: true,
    historyIndex: 0,
  )
  let registry = newHandlerRegistry()

  var currentAction = ContinueReading
  while currentAction == ContinueReading:
    let input = readRawInput()
    if input.len > 0:
      let handler = registry.get(input)
      currentAction = handler(input, prompt, state)
    flushStdout()
  return prompt.content

File: src/io/prompt_handler.rs
use std::io::Write;
use crate::core::session_state::SessionState;
use crate::core::get_prompt::get_prompt;
use crate::models::prompt_state::PromptState;
use crate::io::handler_registry::{HandlerRegistry, InputAction};
use crate::io::read_raw_input::read_raw_input;

pub fn handle_input(state: &mut SessionState) -> String {
    let mut prompt = PromptState {
        prompt: get_prompt(state),
        content: String::new(),
        cursor_position: 0,
        continue_reading: true,
        history_index: 0,
    };
    let handler_registry = HandlerRegistry::new();

    let mut current_action = InputAction::ContinueReading;
    while current_action == InputAction::ContinueReading {
        if let Some(input) = read_raw_input() {
            let handler = handler_registry.get(&input);
            current_action = handler(&input, &mut prompt, state);
        }
        std::io::stdout().flush().ok();
    }
    prompt.content
}
